scripts/src/java/org/oppia/android/scripts/common/BazelClient.kt

```diff
@@ -133,38 +133,38 @@ class BazelClient(private val rootDirectory: File, private val commandExecutor:
   /**
    * Runs code coverage for the specified Bazel test target.
    *
-   * Null return typically occurs when the coverage command fails to generate the 'coverage.dat' file
-   * This can happen due to: Test failures or misconfigurations that prevent the coverage data
-   * from being generated properly.
+   * An empty list being returned typically occurs when the coverage command fails to generate any
+   * 'coverage.dat' file. This can happen due to tests failures or a misconfiguration that prevents
+   * the coverage data from being properly generated.
    *
    * @param bazelTestTarget Bazel test target for which code coverage will be run
-   * @return the generated coverage data as a list of strings
-   *     or null if the coverage data file could not be parsed
+   * @return the generated coverage data as a list of list of strings (since there may be more than
+   *     one file corresponding to a single test target, e.g. in the case of a sharded test), or an
+   *     empty list if no coverage data was found while running the test
    */
-  fun runCoverageForTestTarget(bazelTestTarget: String): List<String>? {
+  fun runCoverageForTestTarget(bazelTestTarget: String): List<List<String>> {
     val instrumentation = bazelTestTarget.split(":")[0]
     val computeInstrumentation = instrumentation.split("/").let { "//${it[2]}/..." }
     val coverageCommandOutputLines = executeBazelCommand(
       "coverage",
       bazelTestTarget,
       "--instrumentation_filter=$computeInstrumentation"
     )
-    return parseCoverageDataFilePath(coverageCommandOutputLines)?.let { path ->
+    return parseCoverageDataFilePath(bazelTestTarget, coverageCommandOutputLines).map { path ->
       File(path).readLines()
     }
   }
 
-  private fun parseCoverageDataFilePath(coverageCommandOutputLines: List<String>): String? {
-    val regex = ".*coverage\\.dat$".toRegex()
-    for (line in coverageCommandOutputLines) {
-      val match = regex.find(line)
-      val extractedPath = match?.value?.substringAfterLast(",")?.trim()
-      if (extractedPath != null) {
-        println("Raw Coverage Data: $extractedPath")
-        return extractedPath
-      }
-    }
-    return null
+  private fun parseCoverageDataFilePath(
+    bazelTestTarget: String,
+    coverageCommandOutputLines: List<String>
+  ): List<String> {
+    // TODO: RD to upstream this & update tests accordingly.
```

Comment:
@Rd4dev can you please bring this into one of your PRs & add corresponding tests?

This properly fixes the fact that sharded tests can produce multiple coverage.dat files.

This should include all the changes in this PR in BazelClient.kt, BazelClientTest.kt, CoverageRunner.kt, CoverageRunnerTest.kt, and RunCoverage.kt (except for the exemption filter update since that's specific to this PR).

Reply:
Woah, yeah! Thanks, Ben! I'll include the changes with the list of files in the PR 2.2 and update here once it's done.

Reply:
Thanks for the fixes @BenHenning, updated the PR 2.2 with all the modifications.

Links to specific parts: [CoverageRunner.kt](https://github.com/oppia/oppia-android/pull/5461/files#diff-e1015bdd1aafa12e3e95de3b408525733c702992881ece1d2399b57ded6a44fdR42), [RunCoverage.kt](https://github.com/oppia/oppia-android/pull/5461/files#diff-62c81482e7c4b720da2d4efc73a48d10369eee097e0bf3439def6648ff150c23R213), [BazelClient.kt](https://github.com/oppia/oppia-android/pull/5461/files#diff-1e798d9f6f6e6059ff0d747cc961c02e0ba9bf4bf588dfe9db710be512d0c63dR158) and added tests for the same - [BazelClientTest.kt - Shard Configured Tests](https://github.com/oppia/oppia-android/pull/5461/files#diff-e50d7ab1968e374641f7298e27dca62ec182a42dc06dac4c47857cd7526de00fR463), [RunCoverageTest.kt - Missing Test Declaration](https://github.com/oppia/oppia-android/pull/5461/files#diff-5a779ea4a8aa2caf5249ced27465f76b8ac19bed5cf250475f6033dfad464908R413)